coordinator/src/main/scala/filodb.coordinator/DatasetCoordinatorActor.scala

```diff
@@ -0,0 +1,237 @@
+   */
+  case class StartFlush(replyTo: Option[ActorRef] = None)
+
+  // Checks if memtable is ready to accept more rows.  Common reason why not is due to lack of memory
+  // or the row limit has been reached
+  case object CanIngest
+
+  /**
+   * Clears all data from the projection.  Waits for existing flush to finish first.
+   */
+  case class ClearProjection(replyTo: ActorRef, projection: Projection)
+
+  /**
+   * Returns current stats.  Note: numRows* will return -1 if the memtable is not set up yet
+   */
+  case object GetStats
+  case class Stats(flushesStarted: Int,
+                   flushesSucceeded: Int,
+                   flushesFailed: Int,
+                   numRowsActive: Int,
+                   numRowsFlushing: Int)
+
+  // Internal messages
+  case class FlushDone(result: Seq[String])
+  case class FlushFailed(t: Throwable)
+
+  def props[K](projection: RichProjection[K],
+               version: Int,
+               columnStore: ColumnStore,
+               reprojector: Reprojector,
+               config: Config): Props =
+    Props(classOf[DatasetCoordinatorActor[K]], projection, version, columnStore, reprojector, config)
+}
+
+/**
+ * The DatasetCoordinatorActor coordinates row ingestion and scheduling reprojections to the columnstore.
+ *
+ * One per (dataset, version) per node.
+ *
+ * Responsible for:
+ * - Owning and setting up MemTables for ingestion
+ * - Feeding rows into the MemTable, and knowing when not to feed the memTable
+ * - Forwarding acks from MemTable back to client
+ * - Scheduling memtable flushes, “flipping” MemTables, and
+ *   calling Reprojector to convert MemTable rows to Segments and flush them to disk
+ *
+ * Scheduling is reactive and not done on a schedule right now.  Flushes are checked and initiated after
+ * rows are inserted and after flushes are done.
+ *
+ * There are usually two MemTables held by this actor.  One, the activeMemTable, ingests incoming rows.
+ * The other, the flushingMemTable, is immutable and holds rows for flushing.  The two are switched.
+ *
+ * ==Configuration==
+ * {{{
+ *   memtable {
+ *     flush-trigger-rows = 50000    # No of rows above which memtable flush might be triggered
+ *     max-rows-per-table = 1000000
+ *     min-free-mb = 512
+ *   }
+ * }}}
+ */
+class DatasetCoordinatorActor[K](projection: RichProjection[K],
+                                 version: Int,
+                                 columnStore: ColumnStore,
+                                 reprojector: Reprojector,
+                                 config: Config) extends BaseActor {
+  import DatasetCoordinatorActor._
+  import context.dispatcher
+
+  val datasetName = projection.dataset.name
+  val nameVer = s"$datasetName/$version"
+  val flushTriggerRows = config.getLong("memtable.flush-trigger-rows")
+  val maxRowsPerTable = config.getInt("memtable.max-rows-per-table")
+  val minFreeMb = config.as[Option[Int]]("memtable.min-free-mb").getOrElse(DefaultMinFreeMb)
```

Comment:
Looks like later on we could consider adding a FiloDBCoreSettings class that would have all of these created once in load time vs n-times in each actor needing the settings.